### ISSUE TITLE
libsubprocess: how to ensure subprocesses are really gone

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1926]
Is there a good way for an exiting comms module to definitively destroy any subprocesses it may have created through the broker exec service?

It is a common pattern for a comms module to handle a request to unload by terminating its reactor and calling a stack of destructors before exiting.  Since subprocess gets notifications of process state changes and completion via reactor callbacks, it seems like this is not possible after the reactor has terminated.

One thought would be to add a function that returns a future that is fulfilled when a subprocess has completed.
```c
flux_future_t *flux_subprocess_wait (flux_subprocess_t *p);
```

Once fulfilled, the user could call `flux_subprocess_exit_code()`, `flux_subprocess_signaled()`, etc, like in the completion callback.  This could work both synchronously, and asynchronously, and could be done under a timeout (in case a subprocess needs a strong urging to terminate).

Any thoughts on the best way to handle this?

In #1922, I resorted to sending a `cmb.disconnect` message to make sure the broker exec service knew to destroy my state.  If I simply closed stdin on my subprocesses (which triggers their exit), there would still be state on exit reported by valgrind.

**Comment by grondo** [comment]
There is already a completion call back to handle an asynchronous wait,
what if flux_subprocess_wait() just blocked by default? Just an idea...

On Jan 16, 2019 9:55 AM, "Jim Garlick" <notifications@github.com> wrote:

Is there a good way for an exiting comms module to definitively destroy any
subprocesses it may have created through the broker exec service?

It is a common pattern for a comms module to handle a request to unload by
terminating its reactor and calling a stack of destructors before exiting.
Since subprocess gets notifications of process state changes and completion
via reactor callbacks, it seems like this is not possible after the reactor
has terminated.

One thought would be to add a function that returns a future that is
fulfilled when a subprocess has completed.

flux_future_t *flux_subprocess_wait (flux_subprocess_t *p);

Once fulfilled, the user could call flux_subprocess_exit_code(),
flux_subprocess_signaled(), etc, like in the completion callback. This
could work both synchronously, and asynchronously, and could be done under
a timeout (in case a subprocess needs a strong urging to terminate).

Any thoughts on the best way to handle this?

In #1922 <https://github.com/flux-framework/flux-core/pull/1922>, I
resorted to sending a cmb.disconnect message to make sure the broker exec
service knew to destroy my state. If I simply closed stdin on my
subprocesses (which triggers their exit), there would still be state on
exit reported by valgrind.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/flux-framework/flux-core/issues/1926>, or mute the
thread

**Comment by garlick** [comment]
> There is already a completion call back to handle an asynchronous wait,
what if flux_subprocess_wait() just blocked by default? Just an idea...

I think that would be sufficient for my case.

**Comment by grondo** [comment]
> In lieu of that, a flux_subprocess_wait() can be added to effectively do something like:

Yeah, I was thinking something similar, though I think `flux_subprocess_wait()` could probably always be blocking since there is already an asynchronous way to wait for processes. So, more simply in @garlick's case

```
void im_unloading (...) {
    flux_subprocess_t *p;
    for p in my_subprocess_list {
          flux_subprocess_wait (p);
          flux_subprocess_destroy (p);
    }
}
```

under the covers a future might be used though.

**Comment by chu11** [comment]
I think there's a subtlety I'm missing:

```
void im_unloading (...) {
    flux_subprocess_t *p;
    for p in my_subprocess_list {
          flux_subprocess_wait (p);
          flux_subprocess_destroy (p);
    }
}
```

Isn't ```im_unloading()``` being called from a callback from the reactor?  So how can we re-enter the reactor in ```flux_subprocess_wait()```.  Seems unsafe to recursively do so?  Perhaps there's some reactor trick I'm unaware of.

**Comment by garlick** [comment]
In my use case, I want to call this from a destructor called after reactor exit.  I don't think re-entering the reactor would be safe there, as it would mean other reactor watchers might still be active (what if timer fires after some destructors have already run, and its handler gets a use-after-free?).

**Comment by grondo** [comment]
>  Isn't im_unloading() being called from a callback from the reactor? So how can we re-enter the reactor in flux_subprocess_wait(). Seems unsafe to recursively do so? Perhaps there's some reactor trick I'm unaware of.

Yeah, I'd say if we add `flux_subprocess_wait (p)` it should only be allowed to be called outside of "main" reactor callbacks. Maybe the function could assert this and error out if not (see `ev_depth`, do we wrap that call in flux?)

**Comment by chu11** [comment]
Ok, one of the things I was confused about was the fact that ```flux_subprocess_wait()``` would presumably have to "wait" for a subprocess to complete.  But to be notified that a subprocess has completed requires that internal callbacks be called (i.e. rpc response), which requires the reactor to still be running.  But the main reactor has already exited.  Thus my confusion.

So it sounds like ```flux_subprocess_wait()```, is more of a ```flux_subprocess_kill_emergency()``` or something like that, that is asking to kill a subprocess outside of the reactor?

**Comment by grondo** [comment]
> But the main reactor has already exited. Thus my confusion.

The main reactor has been forcibly exited, presumably if the subprocess handle `p` is still active, then its callbacks are still active on the main reactor, and if you re-enter the reactor then the RPCs could finish and signal `flux_subprocess_wait (p)` that the subprocess was finally complete.

The problem with this stale timers referencing destructed objects seems like it would be a more generic programming error (destruction of an object referenced by a timer should stop and destroy that timer I would hope!). However, part of the problem is perhaps that a caller wouldn't realize that `flux_subprocess_wait()` is re-entering the reactor.

There is perhaps a larger problem here generally with mixing synchronous and asynchronous code the way we're proposing here. Perhaps the only sane fix is to not exit the reactor until all subprocesses have been destroyed? On a shutdown request, send kill signals and wait for proper completion...

**Comment by garlick** [comment]
> Perhaps the only sane fix is to not exit the reactor until all subprocesses have been destroyed? On a shutdown request, send kill signals and wait for proper completion...

Yeah, I think you're right, now that you put it this way.  Sorry it took me a while to come around!

The other bit is probably the subprocess _service_ shouldn't leak memory if it's being destroyed with subprocesses running or in the process of being reaped.  Maybe we should change the title of this bug to reflect that?

**Comment by chu11** [comment]
Ok, one of the things I was confused about was the fact that ```flux_subprocess_wait()``` would presumably have to "wait" for a subprocess to complete.  But to be notified that a subprocess has completed requires that internal callbacks be called (i.e. rpc response), which requires the reactor to still be running.  But the main reactor has already exited.  Thus my confusion.

So it sounds like ```flux_subprocess_wait()```, is more of a ```flux_subprocess_kill_emergency()``` or something like that, that is asking to kill a subprocess outside of the reactor?

**Comment by grondo** [comment]
> Maybe we should change the title of this bug to reflect that?

I'd say open an issue on that one. It would be cool if we could make a reproducer (does `flux start sh -c 'flux exec sleep 5&'` or something cause the leak?) And add that to the valgrind test...

**Comment by garlick** [comment]
OK I'll open a new issue then.
